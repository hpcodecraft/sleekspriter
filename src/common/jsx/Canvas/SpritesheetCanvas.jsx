var SpritesheetCanvas = React.createClass({
  propTypes: {
    zoom: React.PropTypes.number.isRequired, // zoom level
    file: React.PropTypes.object.isRequired, // FileStore
    pixels: React.PropTypes.object.isRequired, // PixelStore
    maxSize: React.PropTypes.number,
    background: React.PropTypes.string,
  },
  mixins: [CanvasMixin],
  render: function() {

    if(_.isUndefined(this.props.maxSize)) {
      var width = this.props.file.size.width * this.props.file.frames.x * this.props.zoom,
          height = this.props.file.size.height * this.props.file.frames.y * this.props.zoom,
          style = {
            width: width,
            height: height,
          };
    }
    else {
      var fitted = this.fitToSize(this.props.maxSize),
          width = fitted.size.width,
          height = fitted.size.height,
          style = fitted.style;
    }

    return (
      <canvas width={width} height={height} style={style}></canvas>
    );
  },

  paint: function() {

    var canvas = ReactDOM.findDOMNode(this),
        pixels = [],
        layerDict = [];

    flux.stores.FileStore.getData().layers.forEach(function(layer) {
      layerDict[layer.id] = {visible: layer.visible, opacity: layer.opacity};
    }, this);

    this.props.pixels.file.forEach(function(px) {
      if(!pixels[px.z]) pixels[px.z] = [];
      pixels[px.z].push(px);
    });

    this.clear();

    if(this.props.background) {
      var ctx = canvas.getContext('2d');
      ctx.fillStyle = this.props.background;
      ctx.fillRect(0, 0, canvas.width, canvas.height);
    }

    // paint
    pixels.forEach(function(zLayer) {
      zLayer.forEach(function(px) {
        if(layerDict[px.layer].visible === true) {
          var targetPos = this.getPixelSpritesheetPosition(px),
              alpha = px.a * (layerDict[px.layer].opacity / 100);
          Pixel.paint(canvas, targetPos.x, targetPos.y, px.toHex(), alpha, this.props.zoom);
        }
      }, this);
    }, this);
  },

  paintPixel: function() {
    var layerDict = [],
        px = stateHistory.last.payload,
        canvas = ReactDOM.findDOMNode(this);

    flux.stores.FileStore.getData().layers.forEach(function(layer) {
      layerDict[layer.id] = {visible: layer.visible, opacity: layer.opacity};
    }, this);

    if(layerDict[px.layer].visible === true) {
      var targetPos = this.getPixelSpritesheetPosition(px),
          alpha = px.a * (layerDict[px.layer].opacity / 100);
      Pixel.paint(canvas, targetPos.x, targetPos.y, px.color, alpha, this.props.zoom);
    }
  },

  erasePixel: function() {
    // console.log(stateHistory.last.payload);
    this.paint(); // TODO implement a better method instead of repainting everything
  },

  getPixelSpritesheetPosition: function(pixel) {
    var framePos = {
      x: (x = pixel.frame % this.props.file.frames.x) === 0 ? this.props.file.frames.x : x,
      y: Math.ceil(pixel.frame / this.props.file.frames.x),
    };

    var targetPos = {
      x: ((framePos.x-1) * this.props.file.size.width) + pixel.x,
      y: ((framePos.y-1) * this.props.file.size.height) + pixel.y,
    };
    return targetPos;
  },
});
