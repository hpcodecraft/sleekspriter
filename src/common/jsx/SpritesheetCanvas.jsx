var SpritesheetCanvas = React.createClass({
  render: function()Â {
    return (
      <canvas></canvas>
    );
  }
});