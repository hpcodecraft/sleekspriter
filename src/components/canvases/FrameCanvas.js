import React from "react";
import PropTypes from "prop-types";
import { assoc, get } from "sprout-data";
import { CanvasDecorator } from "../decorators";
import { flattenPixels } from "../../utils";
import { sizeShape } from "../../shapes";

class FrameCanvas extends React.Component {
  constructor(props) {
    super(props);

    this.hasMaxSize = this.props.maxSize !== undefined;
    this.hasNoMargin = this.props.noMargin !== undefined;
  }

  render() {
    let width, height, style;

    if (!this.hasMaxSize) {
      width = this.props.size.width * this.props.zoom;
      height = this.props.size.height * this.props.zoom;
      style = {
        width: width,
        height: height,
      };
    } else {
      const fitted = this.props.fitToSize(this.props.maxSize, this.hasNoMargin);
      width = fitted.size.width;
      height = fitted.size.height;
      style = fitted.style;
    }

    return (
      <canvas
        ref={n => (this.canvas = n)}
        width={width}
        height={height}
        style={style}
      />
    );
  }

  componentDidMount() {
    // console.log("FrameCanvas mount!", this.props);
    this.paint();
  }

  componentDidUpdate() {
    // console.log("FrameCanvas update!");
    this.paint();
  }

  mapLayers() {
    // cache all layer z values by layer id
    const zMap = {};

    // cache all layer visible values
    const vMap = {};

    if (this.props.layers) {
      this.props.layers.forEach(layer => {
        zMap[layer.id] = layer.z;
        vMap[layer.id] = layer.visible;
      });
    }

    return {
      zMap,
      vMap,
    };
  }

  paint() {
    // console.log("FrameCanvas paint!", this.canvas, this.props.pixels);
    this.props.clear();

    if (this.props.background) {
      const canvas = this.canvas;
      const ctx = canvas.getContext("2d");
      ctx.fillStyle = this.props.background;
      ctx.fillRect(0, 0, canvas.width, canvas.height);
    }

    if (undefined !== this.props.pixels) {
      const pixels = this.mapPixels();
      const xValues = Object.keys(pixels);
      xValues.map(x => {
        const yValues = Object.keys(pixels[x]);
        yValues.map(y => {
          const p = pixels[x][y][0];
          this.props.paintSinglePixel(
            this.canvas,
            this.props.size,
            x,
            y,
            p.toHex()
          );
        });
      });
    }
  }

  insertPixelInZOrder(zMap, pixels, pixel) {
    // find index to put pixel
    const findNextHigherZ = p => zMap[p.layer] > zMap[pixel.layer];
    const index = pixels.findIndex(findNextHigherZ);

    // insert pixel
    if (index === -1) {
      pixels.unshift(pixel);
    } else {
      pixels.splice(index + 1, 0, pixel);
    }

    return pixels;
  }

  mapPixels() {
    const { zMap, vMap } = this.mapLayers();

    // loop over all flattened pixels and create a map
    // containing the an array of pixels for each x/y-coordinate
    // the array contains all pixels z-sorted, by highest z first
    let pMap = {};

    flattenPixels(this.props.pixels || {}).forEach(pixel => {
      // if layer is visible
      if (vMap[pixel.layer] === true) {
        let pixels = get(pMap, [pixel.x, pixel.y], []);

        if (pixels.length === 0) pixels = [pixel];
        else {
          pixels = this.insertPixelInZOrder(zMap, pixels, pixel);
        }

        pMap = assoc(pMap, [pixel.x, pixel.y], pixels);
      }
    });

    return pMap;
  }

  paintPixel(pixel) {
    const { zMap, vMap } = this.mapLayers();

    const { x, y } = pixel;
    const pMap = this.mapPixels();

    // if layer is visible
    if (vMap[pixel.layer] === true) {
      let pixels = get(pMap, [x, y], []);

      if (pixels.length === 0)
        this.props.paintSinglePixel(
          this.canvas,
          this.props.size,
          x,
          y,
          pixel.toHex()
        );
      else {
        pixels = this.insertPixelInZOrder(zMap, pixels, pixel);
        const p = pixels[0];

        this.props.paintSinglePixel(
          this.canvas,
          this.props.size,
          p.x,
          p.y,
          p.toHex()
        );
      }
    }
  }

  clearPixel(pixel) {
    // console.log("clear", pixel);
    const { vMap } = this.mapLayers();

    const { x, y } = pixel;
    const pMap = this.mapPixels();

    // if layer is visible
    if (vMap[pixel.layer] === true) {
      const pixels = get(pMap, [x, y], []);

      if (pixels.length > 0) {
        const findPixel = p => p.layer === pixel.layer;
        const index = pixels.findIndex(findPixel);

        pixels.splice(index, 1);

        if (pixels.length === 0) {
          this.props.clearSinglePixel(this.canvas, this.props.size, x, y);
        } else {
          const p = pixels[0];

          this.props.paintSinglePixel(
            this.canvas,
            this.props.size,
            p.x,
            p.y,
            p.toHex()
          );
        }
      }
    }
  }
}

FrameCanvas.propTypes = {
  background: PropTypes.string,
  clear: PropTypes.func.isRequired,
  clearSinglePixel: PropTypes.func.isRequired,
  fitToSize: PropTypes.func.isRequired,
  layers: PropTypes.array,
  maxSize: PropTypes.number,
  noMargin: PropTypes.bool,
  paintSinglePixel: PropTypes.func.isRequired,
  pixels: PropTypes.object,
  size: sizeShape.isRequired,
  zoom: PropTypes.number,
};

export { FrameCanvas };
export default CanvasDecorator(FrameCanvas);
