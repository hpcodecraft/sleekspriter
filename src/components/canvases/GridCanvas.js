import React from "react";
import PropTypes from "prop-types";
import { CanvasDecorator } from "../decorators";

class GridCanvas extends React.Component {
  render() {
    const css = {
      width: this.props.width,
      height: this.props.height,
    };

    return (
      <canvas
        ref={n => (this.canvas = n)}
        className="GridCanvas"
        style={css}
        width={this.props.width}
        height={this.props.height}
      />
    );
  }

  componentDidMount() {
    this.drawGrid();
  }

  componentDidUpdate() {
    this.props.clear();
    this.drawGrid();
  }

  drawGrid() {
    const canvas = this.canvas;
    const ctx = canvas.getContext("2d");
    const cell = {
      width: this.props.width / this.props.columns,
      height: this.props.height / this.props.rows,
    };

    if (cell.width >= 2 && cell.height >= 2) {
      ctx.strokeStyle = "#828282";
      ctx.beginPath();

      if (this.props.columns > 1) {
        // vertical lines
        for (let x = cell.width + 0.5; x < this.props.width; x += cell.width) {
          ctx.moveTo(x, 0);
          ctx.lineTo(x, canvas.height);
        }
      }

      if (this.props.rows > 1) {
        // horizontal lines
        for (
          let y = cell.height + 0.5;
          y < this.props.height;
          y += cell.height
        ) {
          ctx.moveTo(0, y);
          ctx.lineTo(canvas.width, y);
        }
      }

      ctx.stroke();
    }
  }
}

GridCanvas.propTypes = {
  clear: PropTypes.func.isRequired,
  columns: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  rows: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
};

export { GridCanvas };
export default CanvasDecorator(GridCanvas);
