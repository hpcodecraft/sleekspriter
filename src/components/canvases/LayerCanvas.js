import React from "react";
import PropTypes from "prop-types";

import { Color } from "../../classes";
import { CanvasDecorator } from "../decorators";

class LayerCanvas extends React.Component {
  constructor(props) {
    super(props);

    this.hasMaxSize = this.props.maxSize !== undefined;
    this.hasZoom = this.props.zoom !== undefined;
  }

  render() {
    let width, height, style;

    if (this.hasMaxSize) {
      const fitted = this.props.fitToSize(this.props.maxSize);
      width = fitted.size.width;
      height = fitted.size.height;
      style = fitted.style;
    } else if (this.hasZoom) {
      width = this.props.size.width * this.props.zoom;
      height = this.props.size.height * this.props.zoom;
      style = {
        width: width,
        height: height,
      };
    }

    return (
      <canvas
        ref={n => (this.canvas = n)}
        width={width}
        height={height}
        style={style}
      />
    );
  }

  componentDidMount() {
    this.paint();
  }

  componentDidUpdate() {
    this.paint();
  }

  paint() {
    this.props.clear();

    if (undefined !== this.props.pixels) {
      const xValues = Object.keys(this.props.pixels);
      xValues.map(x => {
        const yValues = Object.keys(this.props.pixels[x]);
        yValues.map(y => {
          const p = this.props.pixels[x][y];
          const hex = new Color({ rgb: [p.r, p.g, p.b] }).hex();
          this.props.paintSinglePixel(this.canvas, this.props.size, x, y, hex);
        });
      });
    }
  }

  paintPixel(pixel) {
    this.props.paintSinglePixel(
      this.canvas,
      this.props.size,
      pixel.x,
      pixel.y,
      pixel.toHex()
    );
  }

  clearPixel({ x, y }) {
    this.props.clearSinglePixel(this.canvas, this.props.size, x, y);
  }
}

LayerCanvas.propTypes = {
  clear: PropTypes.func.isRequired,
  clearSinglePixel: PropTypes.func.isRequired,
  fitToSize: PropTypes.func.isRequired,
  maxSize: PropTypes.number,
  paintSinglePixel: PropTypes.func.isRequired,
  pixels: PropTypes.object,
  size: PropTypes.object.isRequired, // { width: x, height: y }
  zoom: PropTypes.number,
};

export { LayerCanvas };
export default CanvasDecorator(LayerCanvas);
