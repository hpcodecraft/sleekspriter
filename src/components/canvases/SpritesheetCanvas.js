import React from "react";
import PropTypes from "prop-types";
import { Color } from "../../classes";
import { CanvasDecorator } from "../decorators";

class SpritesheetCanvas extends React.Component {
  constructor(props) {
    super(props);

    this.hasMaxSize = this.props.maxSize != undefined;
  }

  render() {
    let width, height, style;
    const { frames, size, zoom, maxSize, fitToSize } = this.props;

    if(!this.hasMaxSize) {
      width = frames.x * size.width * zoom;
      height = frames.y * size.height * zoom;
      style = {
        width: width,
        height: height,
      };
    }
    else {
      const fitted = fitToSize(maxSize);
      width = fitted.size.width;
      height = fitted.size.height;
      style = fitted.style;
    }

    return <canvas ref="canvas" width={width} height={height} style={style}></canvas>;
  }

  componentDidMount() {
    this.paint();
  }

  componentDidUpdate() {
    this.paint();
  }

  paint() {
    this.props.clear();

    const
      { background } = this.props,
      canvas = this.refs.canvas,
      pixels = [],
      layerDict = [];

    if(background) {
      const ctx = canvas.getContext("2d");
      ctx.fillStyle = background;
      ctx.fillRect(0, 0, canvas.width, canvas.height);
    }

    this.props.layers.forEach(function(layer) {
      layerDict[layer.id] = { visible: layer.visible, opacity: layer.opacity, z: layer.z };
    }, this);

    // sort pixels via z value
    let
      dict = this.props.pixels,
      flen, llen, xlen, ylen,
      frames, f, frame,
      layers, l, layer,
      xValues, x, xValue,
      yValues, y, yValue,
      zValue, pixel;

    frames = Object.keys(dict);
    flen = frames.length;

    for(f = 0; f < flen; f++) {
      frame = frames[f];

      layers = Object.keys(dict[frame]);
      llen = layers.length;

      for(l = 0; l < llen; l++) {
        layer = layers[l];

        zValue = layerDict[layer].z;

        xValues = Object.keys(dict[frame][layer]);
        xlen = xValues.length;

        for(x = 0; x < xlen; x++) {
          xValue = xValues[x];

          yValues = Object.keys(dict[frame][layer][xValue]);
          ylen = yValues.length;

          for(y = 0; y < ylen; y++) {
            yValue = yValues[y];
            pixel = Object.assign({}, dict[frame][layer][xValue][yValue]);

            if(!pixels[zValue]) pixels[zValue] = [];
            pixels[zValue].push(pixel);
          }
        }
      }
    }

    // paint
    pixels.forEach(function(zLayer) {
      zLayer.forEach(function(px) {
        if(layerDict[px.layer].visible === true) {
          const
            targetPos = this.getPixelSpritesheetPosition(px),
            // alpha = px.a * (layerDict[px.layer].opacity / 100), // TODO integrate layer alpha support
            hex = new Color({rgb: [px.r, px.g, px.b]}).hex(),
            sheetSize = this.getSpriteSheetSize(this.props.size, this.props.frames);

          this.props.paintSinglePixel(canvas, sheetSize, targetPos.x, targetPos.y, hex);

          // Pixel.paint(canvas, targetPos.x, targetPos.y, px.toHex(), alpha, zoom);
        }
      }, this);
    }, this);
  }

  getPixelSpritesheetPosition(pixel) {
    const { frames, size } = this.props;

    const x = pixel.frame % frames.x;
    const framePos = {
      x: x === 0 ? frames.x : x,
      y: Math.ceil(pixel.frame / frames.x),
    };

    const targetPos = {
      x: ((framePos.x-1) * size.width) + pixel.x,
      y: ((framePos.y-1) * size.height) + pixel.y,
    };

    return targetPos;
  }

  getSpriteSheetSize(size, frames) {
    return {
      width: size.width * frames.x,
      height: size.height * frames.y
    };
  }
}

SpritesheetCanvas.propTypes = {
  background: PropTypes.string,
  frames: PropTypes.object.isRequired, // { x, y }
  layers: PropTypes.array.isRequired, // array of file layers
  maxSize: PropTypes.number,
  size: PropTypes.object.isRequired, // { width, height }
  zoom: PropTypes.number,
};

export default CanvasDecorator(SpritesheetCanvas);
