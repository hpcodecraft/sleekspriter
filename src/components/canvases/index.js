export FrameCanvas from "./FrameCanvas";
export GridCanvas from "./GridCanvas";
export LayerCanvas from "./LayerCanvas";
export SpritesheetCanvas from "./SpritesheetCanvas";
