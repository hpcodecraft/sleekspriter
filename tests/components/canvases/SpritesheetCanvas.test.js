import { SpritesheetCanvas } from "components/canvases/SpritesheetCanvas";

describe("SpritesheetCanvas", () => {
  let props, wrapper, component, paint;
  beforeEach(() => {
    props = {
      clear: jest.fn(),
      fitToSize: jest.fn(),
      frames: { x: 2, y: 2 },
      layers: [],
      paintSinglePixel: jest.fn(),
      size: { width: 16, height: 16 },
      zoom: 10,
    };

    paint = jest.fn();

    wrapper = shallow(<SpritesheetCanvas {...props} />, {
      disableLifecycleMethods: true,
    });

    component = wrapper.instance();
    component.paint = paint;
  });

  it("should render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should paint itself after mounting", () => {
    component.componentDidMount();
    expect(paint).toHaveBeenCalledTimes(1);
  });

  it("should paint itself after updating", () => {
    component.componentDidUpdate();
    expect(paint).toHaveBeenCalledTimes(1);
  });

  describe("getPixelSpritesheetPosition", () => {
    it("calculates a pixels position in the spritesheet", () => {
      const pos = component.getPixelSpritesheetPosition({
        frame: 4,
        x: 3,
        y: 3,
      });

      expect(pos).toEqual({ x: 19, y: 19 });
    });
  });

  describe("getSpriteSheetSize", () => {
    it("calculates the total spritesheet size", () => {
      const size = component.getSpriteSheetSize(props.size, props.frames);
      expect(size).toEqual({ width: 32, height: 32 });
    });
  });
});
