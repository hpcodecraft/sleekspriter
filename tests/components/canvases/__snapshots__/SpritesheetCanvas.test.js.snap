// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpritesheetCanvas should render correctly 1`] = `"<canvas width=\\"320\\" height=\\"320\\" style=\\"width:320px;height:320px;\\"></canvas>"`;
